gitlab/resource_gitlab_project_variable_test.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/xanzy/go-gitlab"
+)
+
+func TestAccGitlabProjectVariable_basic(t *testing.T) {
+	var projectVariable gitlab.BuildVariable
+	rString := acctest.RandString(5)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckGitlabProjectVariableDestroy,
+		Steps: []resource.TestStep{
+			// Create a project and variable with default options
+			{
+				Config: testAccGitlabProjectVariableConfig(rString),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabProjectVariableExists("gitlab_project_variable.foo", &projectVariable),
+					testAccCheckGitlabProjectVariableAttributes(&projectVariable, &testAccGitlabProjectVariableExpectedAttributes{
+						Key:   fmt.Sprintf("key-%s", rString),
+						Value: fmt.Sprintf("value-%s", rString),
+					}),
+				),
+			},
+			// Update the project variable to toggle all the values to their inverse
+			{
+				Config: testAccGitlabProjectVariableUpdateConfig(rString),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabProjectVariableExists("gitlab_project_hook.foo", &projectVariable),
+					testAccCheckGitlabProjectVariableAttributes(&projectVariable, &testAccGitlabProjectVariableExpectedAttributes{
+						Key:       fmt.Sprintf("key-%s", rString),
+						Value:     fmt.Sprintf("value-inverse-%s", rString),
+						Protected: true,
+					}),
+				),
+			},
+			// Update the project variable to toggle the options back
+			{
+				Config: testAccGitlabProjectVariableConfig(rString),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabProjectVariableExists("gitlab_project_hook.foo", &projectVariable),
+					testAccCheckGitlabProjectVariableAttributes(&projectVariable, &testAccGitlabProjectVariableExpectedAttributes{
+						Key:       fmt.Sprintf("key-%s", rString),
+						Value:     fmt.Sprintf("value-%s", rString),
+						Protected: false,
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabProjectVariableExists(n string, projectVariable *gitlab.BuildVariable) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", n)
+		}
+
+		repoName := rs.Primary.Attributes["project"]
+		if repoName == "" {
+			return fmt.Errorf("No project ID is set")
+		}
+		key := rs.Primary.Attributes["key"]
+		if key == "" {
+			return fmt.Errorf("No variable key is set")
+		}
+		conn := testAccProvider.Meta().(*gitlab.Client)
+
+		gotVariable, _, err := conn.BuildVariables.GetBuildVariable(repoName, key)
+		if err != nil {
+			return err
+		}
+		*projectVariable = *gotVariable
+		return nil
+	}
+}
+
+type testAccGitlabProjectVariableExpectedAttributes struct {
+	Key       string
+	Value     string
+	Protected bool
+}
+
+func testAccCheckGitlabProjectVariableAttributes(variable *gitlab.BuildVariable, want *testAccGitlabProjectVariableExpectedAttributes) resource.TestCheckFunc {
```

Comment:
Instead of creating this helper method you should just be able to use `resource.TestCheckResourceAttr()`

Reply:
I agree with you... Unfortunately my knowledge of Go is so limited I can't find how to make it work properly 😞